### PR TITLE
clickhouse 1.1.54074 (new formula)

### PR DESCRIPTION
**Pull request by hatarist**
- [v] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [v] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [v] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [v] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Here's an OS X build guide: https://github.com/yandex/ClickHouse/blob/master/doc/build_osx.md
In short, we have to recompile boost using gcc-6, hence the `HOMEBREW_CC`/`HOMEBREW_BUILD_FROM_SOURCE` overriding before `depends_on "boost"`.

### CODE REVIEW

File: Formula/clickhouse.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  depends_on "cmake" => :build
+  depends_on "gcc" => :build
+
+  # We have to force some env vars here to force boost to be built from source
+  ENV["HOMEBREW_CC"] = "gcc-6"
+  ENV["HOMEBREW_LD"] = "gcc-6"
+  ENV["HOMEBREW_CXX"] = "g++-6"
+  ENV["HOMEBREW_BUILD_FROM_SOURCE"] = "1"
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: We can't do this. This and Boost need to be build using Clang.

File: Formula/clickhouse.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  ENV["HOMEBREW_BUILD_FROM_SOURCE"] = "0"
+
+  depends_on "icu4c" => :build
+  depends_on "mysql" => :build
+  depends_on "openssl" => :build
+  depends_on "unixodbc" => :build
+  depends_on "glib" => :build
+  depends_on "libtool" => :build
+  depends_on "gettext" => :build
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Are none of these linked to at runtime?

File: Formula/clickhouse.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  depends_on "readline" => :recommended
+
+  def install
+    ENV["DISABLE_MONGODB"] = "1"
+
+    # Hardcode the version assignment since there's no git repository
+    inreplace "libs/libcommon/src/get_revision_lib.sh", /git.*\n.*/, "echo " + version.to_s[-5..-1]
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Use a Git checkout instead of patching.

File: Formula/clickhouse.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    end
+
+    mkdir "#{var}/clickhouse"
+
+    inreplace "#{buildpath}/dbms/src/Server/config.xml" do |s|
+      s.gsub! "/opt/clickhouse/", "#{var}/clickhouse/"
+      s.gsub! "<!-- <max_open_files>262144</max_open_files> -->", "<max_open_files>262144</max_open_files>"
+    end
+
+    # Copy configuration files
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Don't need this comment.

File: Formula/clickhouse.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      system "make"
+      bin.install "#{buildpath}/build/dbms/src/Server/clickhouse" => "clickhouse-server"
+      bin.install_symlink "clickhouse-server" => "clickhouse-client"
+    end
+
+    mkdir "#{var}/clickhouse"
+
+    inreplace "#{buildpath}/dbms/src/Server/config.xml" do |s|
+      s.gsub! "/opt/clickhouse/", "#{var}/clickhouse/"
+      s.gsub! "<!-- <max_open_files>262144</max_open_files> -->", "<max_open_files>262144</max_open_files>"
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Why isn't this the default?

File: Formula/clickhouse.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  depends_on "readline" => :recommended
+
+  def install
+    ENV["DISABLE_MONGODB"] = "1"
+
+    # Hardcode the version assignment since there's no git repository
+    inreplace "libs/libcommon/src/get_revision_lib.sh", /git.*\n.*/, "echo " + version.to_s[-5..-1]
+
+    mkdir "build"
+    cd "build" do
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: `mkdir "build" do`

File: Formula/clickhouse.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      system "cmake", ".."
+      system "make"
+      bin.install "#{buildpath}/build/dbms/src/Server/clickhouse" => "clickhouse-server"
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Is there no `make install` we can use? If not, please request one from upstream.

File: Formula/clickhouse.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    The configuration files are available at:
+      #{etc}/clickhouse-client/
+      #{etc}/clickhouse-server/
+    The database itself will store data at:
+      #{var}/clickhouse/
+
+    If you're going to run the server, make sure to increase `maxfiles` limit:
+      https://github.com/yandex/ClickHouse/blob/master/MacOS.md
+  EOS
+  end
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Don't need any of these caveats as they aren't Homebrew specific or different to normal.

File: Formula/clickhouse.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+  test do
+    system "#{bin}/clickhouse-client", "--help"
```
Comment (MikeMcQuaid) [inline Formula/clickhouse.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/clickhouse.rb (hunk 9 above)
Comment (hatarist) [inline Formula/clickhouse.rb]: Thanks a lot for taking your time to review!

**Comment by woodruffw** [comment]
Any update here @hatarist?

**Comment by hatarist** [comment]
I'm very sorry, I have little to no experience with cmake/make and it seems that the project has a lot of issues regarding build-scripts. (see [clickhouse issue #220 (ru->en translation)](https://translate.google.com/translate?depth=1&hl=en&ie=UTF8&prev=_t&rurl=translate.google.com&sl=ru&tl=en&u=https://github.com/yandex/ClickHouse/issues/220))
In short, people are blaming the poor code quality of the build-scripts. Cons are: it looks for the static libraries, it doesn't use FindOpenSSL/FindBoost, it doesn't look for termcap/libz libraries (as it's used in the project), it doesn't use cmake -D options and uses environment variables instead.

I also have no idea what to do with
> We can't do this. This and Boost need to be build using Clang.
:(

Sorry again. I guess this pull request should be closed until most of these things could be fixed?

**Comment by woodruffw** [comment]
Don't worry about it! Thanks for making an effort to build this despite its build system difficulties. I'll close for now, but future PR(s) for this formula are always welcome.

Thanks again!
